// @vtex/gatsby-plugin-vtex-nginx/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FILE_SERVE_DIRECTIVE_ENV_VAR = exports.LOCATIONS_ONLY_ENV_VAR = exports.VTEX_NGINX_CONF_FILENAME = exports.CACHING_HEADERS = exports.PAGE_DATA_DIR = exports.COMMON_BUNDLES = exports.PUBLIC_CACHING_HEADER = exports.IMMUTABLE_CACHING_HEADER = exports.INDEX_HTML = exports.BUILD_HTML_STAGE = void 0;
exports.BUILD_HTML_STAGE = 'build-html';
exports.INDEX_HTML = 'index.html';
exports.IMMUTABLE_CACHING_HEADER = {
    name: 'Cache-Control',
    value: 'public, max-age=31536000, immutable',
};
exports.PUBLIC_CACHING_HEADER = {
    name: 'Cache-Control',
    value: 'public, max-age=0, must-revalidate',
};
exports.COMMON_BUNDLES = ['commons', 'app'];
exports.PAGE_DATA_DIR = 'page-data';
exports.CACHING_HEADERS = [
    ['/sw.js', [{ name: 'Cache-Control', value: 'no-cache' }]],
];
exports.VTEX_NGINX_CONF_FILENAME = 'nginx.conf';
exports.LOCATIONS_ONLY_ENV_VAR = 'NGINX_LOCATIONS_ONLY';
exports.FILE_SERVE_DIRECTIVE_ENV_VAR = 'NGINX_SERVE_FILE_DIRECTIVE';
