// @vtex/gatsby-plugin-vtex-nginx/gatsby-node.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.onPostBuild = exports.onCreateWebpackConfig = void 0;
const path_1 = require("path");
const fs_1 = require("fs");
const webpack_assets_manifest_1 = __importDefault(require("webpack-assets-manifest"));
const constants_1 = require("./constants");
const headers_1 = require("./headers");
const nginx_generator_1 = require("./nginx-generator");
const listFiles_1 = require("./listFiles");
const pluginOptions_1 = require("./pluginOptions");
const assetsManifest = {};
const Node = {
    onCreateWebpackConfig({ actions, stage }) {
        if (stage !== constants_1.BUILD_HTML_STAGE) {
            return;
        }
        actions.setWebpackConfig({
            plugins: [
                new webpack_assets_manifest_1.default({
                    assets: assetsManifest,
                    merge: true,
                }),
            ],
        });
    },
    async onPostBuild({ store, pathPrefix, reporter }, opt) {
        const options = pluginOptions_1.pluginOptions(opt);
        const { program, pages: pagesMap, redirects } = store.getState();
        const pages = Array.from(pagesMap.values());
        const rewrites = pages
            .filter((page) => page.matchPath && page.matchPath !== page.path)
            .map((page) => ({
            fromPath: page.matchPath,
            toPath: page.path,
        }));
        const publicFolder = path_1.join(program.directory, 'public');
        const { assetsByChunkName } = require(path_1.join(publicFolder, 'webpack.stats.json'));
        const manifest = {
            ...mapObjectValues(assetsManifest, (value) => [value]),
            ...assetsByChunkName,
        };
        const files = await listFiles_1.listFilesRecursively(publicFolder);
        let headers = {
            ...headers_1.emptyHeadersMapForFiles(files),
            ...headers_1.preloadHeadersByPath(pages, manifest, pathPrefix),
            ...headers_1.cacheHeadersByPath(pages, manifest),
        };
        if (typeof options.transformHeaders === 'function') {
            headers = headers_1.applyUserHeadersTransform(headers, options.transformHeaders);
        }
        headers = headers_1.addStaticCachingHeader(headers);
        headers = headers_1.addPublicCachingHeader(headers);
        fs_1.writeFileSync(path_1.join(program.directory, 'public', constants_1.VTEX_NGINX_CONF_FILENAME), nginx_generator_1.generateNginxConfiguration(rewrites, redirects, headers, files, options));
        reporter.success('write out nginx configuration');
    },
};
function mapObjectValues(obj, transform) {
    return Object.fromEntries(Object.entries(obj).map(([k, v]) => [k, transform(v)]));
}
exports.onCreateWebpackConfig = Node.onCreateWebpackConfig;
exports.onPostBuild = Node.onPostBuild;
