// @vtex/gatsby-plugin-vtex-nginx/pluginOptions.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.pluginOptions = void 0;
const constants_1 = require("./constants");
const defaultOptions = {
    transformHeaders: undefined,
    writeOnlyLocations: false,
    serveFileDirective: ['try_files', '/$file', '=404'],
    plugins: [],
};
function pluginOptions(options) {
    var _a, _b, _c, _d, _e, _f;
    return {
        transformHeaders: (_a = options.transformHeaders) !== null && _a !== void 0 ? _a : defaultOptions.transformHeaders,
        writeOnlyLocations: (_c = (_b = safeTransform(process.env[constants_1.LOCATIONS_ONLY_ENV_VAR], (x) => Boolean(JSON.parse(x)))) !== null && _b !== void 0 ? _b : options.writeOnlyLocations) !== null && _c !== void 0 ? _c : defaultOptions.writeOnlyLocations,
        serveFileDirective: (_e = (_d = safeTransform(process.env[constants_1.FILE_SERVE_DIRECTIVE_ENV_VAR], JSON.parse)) !== null && _d !== void 0 ? _d : options.serveFileDirective) !== null && _e !== void 0 ? _e : defaultOptions.serveFileDirective,
        plugins: (_f = options.plugins) !== null && _f !== void 0 ? _f : defaultOptions.plugins,
    };
}
exports.pluginOptions = pluginOptions;
function safeTransform(value, f) {
    if (value === undefined) {
        return undefined;
    }
    try {
        return f(value);
    }
    catch {
        return undefined;
    }
}
