// @vtex/gatsby-theme-vtex/components/Minicart/Drawer/Content.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderMinicartDrawerContent = void 0;
const react_1 = __importDefault(require("react"));
const theme_ui_1 = require("theme-ui");
const useNumberFormat_1 = require("../../../sdk/localization/useNumberFormat");
const Delete_1 = __importDefault(require("./Delete"));
const Quantity_1 = __importDefault(require("./Quantity"));
exports.HeaderMinicartDrawerContent = ({ data, variant, imageElement, }) => {
    const customVariant = `${variant}.content`;
    const imageProps = (item) => ({
        as: imageElement,
        width: 200,
        height: 200,
        src: item.image.src,
        alt: item.image.alt,
        loading: 'lazy',
    });
    const { format } = useNumberFormat_1.useNumberFormat();
    return (react_1.default.createElement(theme_ui_1.Flex, { variant: customVariant }, data.map((item, idx) => (react_1.default.createElement(theme_ui_1.Flex, { key: item.id, variant: `${customVariant}.product` },
        react_1.default.createElement(theme_ui_1.Box, { variant: `${customVariant}.product.image` },
            react_1.default.createElement(theme_ui_1.Box, Object.assign({}, imageProps(item)))),
        react_1.default.createElement(theme_ui_1.Flex, { variant: `${customVariant}.product.name` },
            react_1.default.createElement(theme_ui_1.Flex, null,
                react_1.default.createElement(theme_ui_1.Text, { variant: `${customVariant}.product.name.text` }, item.name),
                react_1.default.createElement(Delete_1.default, { index: idx, variant: customVariant })),
            react_1.default.createElement(Quantity_1.default, { index: idx, variant: customVariant }),
            react_1.default.createElement(theme_ui_1.Text, { variant: `${customVariant}.product.name.value` }, format(Number(item.price) / 100))))))));
};
