// @vtex/gatsby-theme-vtex/components/Minicart/Drawer/index.jsx
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const store_ui_1 = require("@vtex/store-ui");
const react_1 = __importDefault(require("react"));
const useOrderForm_1 = require("../../../sdk/orderForm/useOrderForm");
const constants_1 = require("../../../sdk/product/constants");
const Content_1 = require("./Content");
const Footer_1 = require("./Footer");
const Header_1 = require("./Header");
const useHeaderMinicartDrawerContentData = (orderForm) => {
    var _a;
    const data = (_a = orderForm.value) === null || _a === void 0 ? void 0 : _a.items.map((item) => {
        var _a, _b;
        return ({
            id: Number(item.uniqueId),
            image: {
                alt: item.name,
                src: (_b = (_a = item.imageUrls) === null || _a === void 0 ? void 0 : _a.at1x) !== null && _b !== void 0 ? _b : constants_1.IMAGE_DEFAULT,
            },
            name: item.name,
            price: item.price,
        });
    });
    return {
        data: data || [],
    };
};
const CustomMinicartDrawer = ({ isOpen, onClose, variant, }) => {
    var _a, _b;
    const orderForm = useOrderForm_1.useOrderForm();
    const count = (_b = (_a = orderForm === null || orderForm === void 0 ? void 0 : orderForm.value) === null || _a === void 0 ? void 0 : _a.items.length) !== null && _b !== void 0 ? _b : 0;
    const contentData = useHeaderMinicartDrawerContentData(orderForm);
    const customVariant = `${variant}.drawer`;
    return (<store_ui_1.MinicartDrawer variant={variant} isOpen={isOpen} placement="right" onClose={onClose} width={400}>
      <Header_1.HeaderMinicartDrawerHeader onClose={onClose} count={count} variant={customVariant}/>
      <Content_1.HeaderMinicartDrawerContent {...contentData} imageElement={store_ui_1.AspectImage} variant={customVariant}/>
      <Footer_1.HeaderMinicartDrawerFooter variant={customVariant}/>
    </store_ui_1.MinicartDrawer>);
};
exports.default = CustomMinicartDrawer;
