// @vtex/lighthouse-config/dist/lighthouse-config.esm.js
var lhConfig = function lhConfig(_ref) {
  var urls = _ref.urls,
      server = _ref.server;
  var url = urls.map(function (path) {
    return "" + server + path;
  });
  return {
    ci: {
      collect: {
        url: url,
        numberOfRuns: 5
      },
      assert: {
        preset: 'lighthouse:no-pwa',
        assertions: {
          'categories:accessibility': ['error', {
            minScore: 0.95
          }],
          'categories:best-practices': ['error', {
            minScore: 0.95
          }],
          'categories:performance': ['error', {
            minScore: 0.95
          }],
          'categories:pwa': 'off',
          'categories:seo': ['error', {
            minScore: 0.9
          }],
          'cumulative-layout-shift': ['error', {
            maxNumericValue: 0.2
          }],
          'first-contentful-paint': ['error', {
            maxNumericValue: 2000
          }],
          'is-crawlable': 'off',
          'largest-contentful-paint': ['error', {
            maxNumericValue: 3500
          }],
          'max-potential-fid': ['error', {
            maxNumericValue: 300
          }],
          'total-blocking-time': ['error', {
            maxNumericValue: 200
          }],
          'unused-javascript': ['error', {
            maxLength: 10
          }],
          'uses-long-cache-ttl': 'off',
          'uses-rel-preconnect': 'warn',
          'uses-text-compression': ['warn', {
            maxLength: 1
          }],
          bypass: 'off',
          interactive: ['error', {
            maxNumericValue: 3000
          }]
        }
      },
      upload: {
        target: 'temporary-public-storage',
        urlReplacementPatterns: 's/.*app//'
      }
    }
  };
};

export default lhConfig;
