// @vtex/gatsby-theme-vtex/templates/product.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.query = void 0;
const react_1 = __importStar(require("react"));
const gatsby_1 = require("gatsby");
const ErrorBoundary_1 = __importDefault(require("../components/Error/ErrorBoundary"));
const HybridWrapper_1 = __importDefault(require("../components/HybridWrapper"));
const Layout_1 = __importDefault(require("../components/Layout"));
const AboveTheFold_1 = __importDefault(require("../components/ProductPage/AboveTheFold"));
const AboveTheFoldPreview_1 = __importDefault(require("../components/ProductPage/AboveTheFoldPreview"));
const BelowTheFoldPreview_1 = __importDefault(require("../components/ProductPage/BelowTheFoldPreview"));
const Viewport_1 = __importDefault(require("../components/Suspense/Viewport"));
const useQuery_1 = require("../sdk/graphql/useQuery");
const ProductPageQuery_graphql_1 = require("./__generated__/ProductPageQuery.graphql");
const SEO_1 = __importDefault(require("../components/ProductPage/SEO"));
const ErrorHandler_1 = __importDefault(require("../components/Error/ErrorHandler"));
const usePixelSendEvent_1 = require("../sdk/pixel/usePixelSendEvent");
const belowTheFoldPreloader = () => Promise.resolve().then(() => __importStar(require('../components/ProductPage/BelowTheFold')));
const BelowTheFold = react_1.lazy(belowTheFoldPreloader);
const ProductPage = (props) => {
    var _a, _b, _c;
    const { data: initialData, pageContext, slug: routeSlug } = props;
    const { staticPath } = pageContext;
    const slug = ((_a = pageContext.slug) !== null && _a !== void 0 ? _a : routeSlug);
    const { data } = useQuery_1.useQuery(Object.assign(Object.assign({}, ProductPageQuery_graphql_1.ProductPageQuery), { variables: { slug, staticPath: true }, suspense: true, initialData: staticPath ? initialData : undefined }));
    usePixelSendEvent_1.usePixelSendEvent(() => [
        {
            type: 'vtex:pageView',
            data: {
                pageUrl: window.location.href,
                pageTitle: document.title,
                referrer: document.referrer,
                accountName: process.env.GATSBY_VTEX_TENANT,
            },
        },
        {
            type: 'vtex:productView',
            data: {
                product: data === null || data === void 0 ? void 0 : data.vtex.product,
            },
        },
    ], (_c = (_b = data === null || data === void 0 ? void 0 : data.vtex.product) === null || _b === void 0 ? void 0 : _b.productId) !== null && _c !== void 0 ? _c : '');
    const pageProps = Object.assign(Object.assign({}, props), { data: data, slug });
    return (react_1.default.createElement(react_1.Fragment, null,
        react_1.default.createElement(AboveTheFold_1.default, Object.assign({}, pageProps)),
        react_1.default.createElement(SEO_1.default, Object.assign({}, pageProps)),
        react_1.default.createElement(Viewport_1.default, { fallback: react_1.default.createElement(BelowTheFoldPreview_1.default, null), preloader: belowTheFoldPreloader },
            react_1.default.createElement(BelowTheFold, Object.assign({}, pageProps)))));
};
const Page = (props) => {
    const { pageContext: { staticPath }, } = props;
    return (react_1.default.createElement(Layout_1.default, null,
        react_1.default.createElement(HybridWrapper_1.default, { isPrerendered: staticPath, fallback: react_1.default.createElement(AboveTheFoldPreview_1.default, null) },
            react_1.default.createElement(ErrorBoundary_1.default, { fallback: (error) => react_1.default.createElement(ErrorHandler_1.default, { error: error }) },
                react_1.default.createElement(ProductPage, Object.assign({}, props))))));
};
exports.query = gatsby_1.graphql `
  query ProductPageQuery($slug: String, $staticPath: Boolean!) {
    vtex {
      product(slug: $slug) @include(if: $staticPath) {
        ...ProductDetailsTemplate_product
        ...StructuredProductFragment_product
        productId
        description
        categoryTree {
          name
          href
        }
        items {
          itemId
        }
      }
    }
  }
`;
exports.default = Page;
