// @vtex/gatsby-theme-vtex/templates/search.jsx
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.query = void 0;
/* eslint-disable no-shadow */
const gatsby_1 = require("gatsby");
const react_1 = __importStar(require("react"));
const ErrorBoundary_1 = __importDefault(require("../components/Error/ErrorBoundary"));
const ErrorHandler_1 = __importDefault(require("../components/Error/ErrorHandler"));
const HybridWrapper_1 = __importDefault(require("../components/HybridWrapper"));
const Layout_1 = __importDefault(require("../components/Layout"));
const AboveTheFold_1 = __importDefault(require("../components/SearchPage/AboveTheFold"));
const AboveTheFoldPreview_1 = __importDefault(require("../components/SearchPage/AboveTheFoldPreview"));
const BelowTheFoldPreview_1 = __importDefault(require("../components/SearchPage/BelowTheFoldPreview"));
const SEO_1 = __importDefault(require("../components/SearchPage/SEO"));
const Viewport_1 = __importDefault(require("../components/Suspense/Viewport"));
const useQuery_1 = require("../sdk/graphql/useQuery");
const Provider_1 = require("../sdk/search/Provider");
const useSearchFiltersFromPageContext_1 = require("../sdk/search/useSearchFiltersFromPageContext");
const useRCSendEvent_1 = require("../sdk/vtexrc/useRCSendEvent");
const SearchPageQuery_graphql_1 = require("./__generated__/SearchPageQuery.graphql");
const belowTheFoldPreloader = () => Promise.resolve().then(() => __importStar(require('../components/SearchPage/BelowTheFold')));
const BelowTheFold = react_1.lazy(belowTheFoldPreloader);
const SearchPage = (props) => {
    var _a, _b;
    const { pageContext, data: staticData } = props;
    const { staticPath } = pageContext;
    const filters = useSearchFiltersFromPageContext_1.useSearchFiltersFromPageContext(pageContext);
    const { data } = useQuery_1.useQuery(Object.assign(Object.assign({}, SearchPageQuery_graphql_1.SearchPageQuery), { variables: Object.assign(Object.assign({}, filters), { staticPath: true }), suspense: true, initialData: staticPath ? staticData : undefined }));
    useRCSendEvent_1.useRCSendEvent({
        type: 'internalSearchView',
        payload: {
            siteSearchForm: filters.query,
            siteSearchTerm: filters.fullText,
            siteSearchResults: (_b = (_a = data === null || data === void 0 ? void 0 : data.vtex.productSearch) === null || _a === void 0 ? void 0 : _a.recordsFiltered) !== null && _b !== void 0 ? _b : 0,
        },
    });
    if (!data) {
        return <div>Not Found</div>;
    }
    return (<Provider_1.SearchProvider filters={filters} data={data}>
      <AboveTheFold_1.default {...props} data={data}/>
      <SEO_1.default {...props} data={data}/>
      <Viewport_1.default fallback={<BelowTheFoldPreview_1.default />} preloader={belowTheFoldPreloader}>
        <BelowTheFold />
      </Viewport_1.default>
    </Provider_1.SearchProvider>);
};
const Page = (props) => {
    const { pageContext: { staticPath }, } = props;
    return (<Layout_1.default>
      <HybridWrapper_1.default isPrerendered={staticPath} fallback={<AboveTheFoldPreview_1.default />}>
        <ErrorBoundary_1.default fallback={(error) => <ErrorHandler_1.default error={error}/>}>
          <SearchPage {...props}/>
        </ErrorBoundary_1.default>
      </HybridWrapper_1.default>
    </Layout_1.default>);
};
exports.query = gatsby_1.graphql `
  query SearchPageQuery(
    $query: String
    $map: String
    $fullText: String
    $staticPath: Boolean!
    $selectedFacets: [VTEX_SelectedFacetInput!]
    $orderBy: String = "OrderByScoreDESC"
  ) {
    vtex {
      productSearch(
        from: 0
        to: 9
        hideUnavailableItems: true
        productOriginVtex: true
        simulationBehavior: skip
        orderBy: $orderBy
        query: $query
        map: $map
        fullText: $fullText
        selectedFacets: $selectedFacets
      ) @include(if: $staticPath) {
        products {
          productId
          productName
          linkText
          items {
            itemId
            images {
              imageUrl
              imageText
            }
          }
        }
        titleTag
        recordsFiltered
      }
      facets(
        query: $query
        map: $map
        fullText: $fullText
        selectedFacets: $selectedFacets
        operator: or
        behavior: "Static"
      ) @include(if: $staticPath) {
        breadcrumb {
          href
          name
        }
        facets {
          name
          type
          values {
            key
            name
            value
            selected
            quantity
            values: children {
              key
              name
              value
              selected
              quantity
              values: children {
                key
                name
                value
                selected
                quantity
              }
            }
          }
        }
      }
    }
  }
`;
exports.default = Page;
