// @vtex/gatsby-theme-vtex/sdk/product/useDetailsImage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useDetailsImage = void 0;
const router_1 = require("@reach/router");
const react_1 = require("react");
const fileManager_1 = require("../img/fileManager");
const constants_1 = require("./constants");
const DEFAULT_IMAGES = [
    {
        imageUrl: constants_1.IMAGE_DEFAULT,
        imageText: 'Product Image',
    },
];
const maxWidth = constants_1.DETAILS_IMAGE[constants_1.DETAILS_IMAGE.length - 1].widthStr;
const maxHeight = constants_1.DETAILS_IMAGE[constants_1.DETAILS_IMAGE.length - 1].heightStr;
const sizes = constants_1.DETAILS_IMAGE.map(({ media }) => media).join(', ');
exports.useDetailsImage = (maybeImages) => {
    const { state } = router_1.useLocation();
    const images = maybeImages !== null && maybeImages !== void 0 ? maybeImages : DEFAULT_IMAGES;
    return react_1.useMemo(() => images.map(({ imageUrl, imageText }, index) => {
        const src = imageUrl !== null && imageUrl !== void 0 ? imageUrl : constants_1.IMAGE_DEFAULT;
        const useSummary = (state === null || state === void 0 ? void 0 : state.fromSummary) && index === 0;
        const srcSet = constants_1.DETAILS_IMAGE.map(({ width, height }) => `${fileManager_1.scaleFileManagerImage(src, width, height)} ${width}w`).join(', ');
        const srcSetPlaceholder = useSummary
            ? fileManager_1.scaleFileManagerImage(src, constants_1.SUMMARY_IMAGE.width, constants_1.SUMMARY_IMAGE.height)
            : srcSet;
        const sizesPlaceholder = useSummary ? '(minWidth: 0px) 100vw' : sizes;
        return {
            targetProps: {
                sizes,
                srcSet,
            },
            placeholderProps: {
                sizes: sizesPlaceholder,
                srcSet: srcSetPlaceholder,
            },
            src: imageUrl,
            alt: imageText,
            loading: 'eager',
            width: maxWidth,
            height: maxHeight,
        };
    }), [images]);
};
