// @vtex/gatsby-theme-vtex/sdk/auth/Service/validateAccessKey.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateAccessKey = void 0;
const api_1 = require("./api");
exports.validateAccessKey = ({ recaptcha = '', setPreference, accessKey, login, }) => __awaiter(void 0, void 0, void 0, function* () {
    const form = new FormData();
    form.append('login', login);
    form.append('accessKey', accessKey);
    form.append('recaptcha', recaptcha);
    const search = new URLSearchParams();
    search.append('setPreference', `${!!setPreference}`);
    const response = yield fetch(`${api_1.api.pub.authentication.accesskey.validate}?${search.toString()}`, {
        method: 'POST',
        credentials: 'include',
        body: form,
        headers: {
            accept: 'application/json',
        },
    });
    if (response.status > 300) {
        throw new Error('Something went wrong while logging in');
    }
});
