// @vtex/gatsby-theme-vtex/sdk/orderForm/controller/addToCart.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.mutation = exports.addToCart = void 0;
const gatsby_plugin_graphql_1 = require("@vtex/gatsby-plugin-graphql");
const request_1 = require("../../graphql/request");
const orderForm_1 = require("./orderForm");
const AddToCartMutation_graphql_1 = require("./__generated__/AddToCartMutation.graphql");
exports.addToCart = (orderFormId, items, cb) => __awaiter(void 0, void 0, void 0, function* () {
    return orderForm_1.queue.add(() => __awaiter(void 0, void 0, void 0, function* () {
        const { addToCart: of } = yield request_1.request(Object.assign(Object.assign({}, AddToCartMutation_graphql_1.AddToCartMutation), { variables: { orderFormId, items } }));
        orderForm_1.setOrderFormState(of, cb);
        return of;
    }));
});
exports.mutation = gatsby_plugin_graphql_1.gql `
  mutation AddToCartMutation(
    $orderFormId: ID
    $items: [VTEX_ItemInput]
    $marketingData: VTEX_MarketingDataInput
  ) {
    addToCart(
      orderFormId: $orderFormId
      items: $items
      marketingData: $marketingData
    ) {
      ...OrderFormFragment_orderForm
    }
  }
`;
