// @vtex/gatsby-plugin-theme-ui/gatsby-node.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.onCreateWebpackConfig = exports.onCreateBabelConfig = exports.onPostBootstrap = void 0;
const path_1 = require("path");
const root = process.cwd();
const name = '@vtex/gatsby-plugin-theme-ui';
const base = path_1.join(root, 'src', name);
const filepath = path_1.join(base, 'index.ts');
exports.onPostBootstrap = () => {
    require('@babel/register')({
        extensions: ['.ts'],
        presets: ['@babel/preset-typescript'],
    });
    const { default: theme } = require(filepath);
    global.theme = theme;
};
exports.onCreateBabelConfig = ({ actions: { setBabelPlugin }, }) => {
    if (process.env.NODE_ENV !== 'production') {
        return;
    }
    setBabelPlugin({
        name: require.resolve('./babel'),
        options: {
            inFile: filepath,
            inPath: base,
        },
    });
};
exports.onCreateWebpackConfig = ({ actions: { setWebpackConfig }, }) => {
    setWebpackConfig({
        resolve: {
            alias: {
                '@emotion/is-prop-valid': require.resolve('./@emotion/is-prop-valid/index.js'),
            },
        },
    });
};
