// @vtex/gatsby-theme-vtex/components/ProductSummary/index.jsx
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.fragment = void 0;
const gatsby_1 = require("gatsby");
const react_1 = __importDefault(require("react"));
const store_ui_1 = require("@vtex/store-ui");
const ProductSummary = () => (<store_ui_1.Flex sx={{ flexWrap: 'wrap' }}>
    <store_ui_1.Center height="50px">
      <store_ui_1.Text>This is the product summary component</store_ui_1.Text>
    </store_ui_1.Center>
    <store_ui_1.Center height="120px">
      <store_ui_1.Text>
        This component is used in many different parts of your store, like in
        the shelf and search results
      </store_ui_1.Text>
    </store_ui_1.Center>
    <store_ui_1.Center height="120px">
      <store_ui_1.Text>
        Base building blocks are available in <strong>@vtex/store-ui</strong>.
        Use them to create your own cusstom version of product summary via
        Shadowing.
      </store_ui_1.Text>
    </store_ui_1.Center>
  </store_ui_1.Flex>);
exports.fragment = gatsby_1.graphql `
  fragment ProductSummary_product on VTEX_Product {
    productId
    productName
    linkText
    productClusters {
      name
    }
    items {
      itemId
      images {
        imageUrl
        imageText
      }
      sellers {
        sellerId
        commercialOffer: commertialOffer {
          maxInstallments: Installments(criteria: MAX_WITHOUT_INTEREST) {
            value: Value
            numberOfInstallments: NumberOfInstallments
          }
          installments: Installments(criteria: ALL) {
            value: Value
            numberOfInstallments: NumberOfInstallments
            interestRate: InterestRate
          }
          availableQuantity: AvailableQuantity
          price: Price
          listPrice: ListPrice
          spotPrice
          teasers {
            name
          }
        }
      }
    }
  }
`;
exports.default = ProductSummary;
