// @vtex/gatsby-theme-vtex/sdk/auth/validate.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isValidAccessCode = exports.isValidPassword = exports.isValidEmail = void 0;
const patterns = {
    hasNumber: /\d/,
    lowerCaseLetter: /[a-z]/,
    upperCaseLetter: /[A-Z]/,
    email: /^(([^<>()[\]\\.,;:\s@!"]+(\.[^<>()[\]\\.,;:\s@!"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/,
};
exports.isValidEmail = (text) => patterns.email.test(text.toLowerCase());
exports.isValidPassword = (text) => {
    const hasMinLength = text.length >= 8;
    const hasNumber = patterns.hasNumber.test(text);
    const hasLowerCaseLetter = patterns.lowerCaseLetter.test(text);
    const hasUpperCaseLetter = patterns.upperCaseLetter.test(text);
    return {
        passwordIsValid: hasNumber && hasMinLength && hasLowerCaseLetter && hasUpperCaseLetter,
        hasNumber,
        hasMinLength,
        hasLowerCaseLetter,
        hasUpperCaseLetter,
    };
};
exports.isValidAccessCode = (text) => text.length === 6;
