// @vtex/gatsby-theme-vtex/pages/account.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const gatsby_plugin_i18n_1 = require("@vtex/gatsby-plugin-i18n");
const store_ui_1 = require("@vtex/store-ui");
const gatsby_1 = require("gatsby");
const react_1 = __importStar(require("react"));
const Container_1 = __importDefault(require("../components/Container"));
const Layout_1 = __importDefault(require("../components/Layout"));
const SSR_1 = __importDefault(require("../components/Suspense/SSR"));
const RenderExtensionLoader_1 = __importDefault(require("../sdk/RenderExtensionLoader"));
const useProfile_1 = require("../sdk/session/useProfile");
const MY_ACCOUNT_PATH = '/account';
const MY_ACCOUNT_DIV_NAME = 'my-account';
const MY_ACCOUNT_EXTENSION_NAME = 'my-account-portal';
const ONE_MIN_IN_MILLI = 60 * 100;
const render = (locale) => __awaiter(void 0, void 0, void 0, function* () {
    const loader = new RenderExtensionLoader_1.default({
        account: process.env.GATSBY_VTEX_TENANT,
        workspace: process.env.GATSBY_VTEX_IO_WORKSPACE,
        verbose: process.env.NODE_ENV !== 'production',
        publicEndpoint: undefined,
        timeout: ONE_MIN_IN_MILLI,
        path: MY_ACCOUNT_PATH,
        locale,
    });
    const myAccountDiv = document.getElementById(MY_ACCOUNT_DIV_NAME);
    if (window.__RENDER_7_RUNTIME__) {
        loader.render(MY_ACCOUNT_EXTENSION_NAME, myAccountDiv, undefined);
        return;
    }
    yield loader.load();
    window.__RUNTIME__ = loader.render(MY_ACCOUNT_EXTENSION_NAME, myAccountDiv, undefined);
});
const Loading = () => (react_1.default.createElement(store_ui_1.Center, { height: "750px" },
    react_1.default.createElement(store_ui_1.Spinner, null)));
const MyAccount = () => {
    var _a;
    const profile = useProfile_1.useProfile({ stale: false });
    const isAuthenticated = ((_a = profile === null || profile === void 0 ? void 0 : profile.isAuthenticated) === null || _a === void 0 ? void 0 : _a.value) === 'true';
    const [loading, setLoading] = react_1.useState(true);
    const { locale, defaultLocale } = gatsby_plugin_i18n_1.useIntl();
    react_1.useEffect(() => {
        ;
        (() => __awaiter(void 0, void 0, void 0, function* () {
            try {
                if (!isAuthenticated) {
                    const path = gatsby_plugin_i18n_1.localizedPath(defaultLocale, locale, '/login');
                    gatsby_1.navigate(path);
                    return;
                }
                yield render(locale);
            }
            catch (err) {
                console.error(err);
            }
            finally {
                setLoading(false);
            }
        }))();
    }, [defaultLocale, isAuthenticated, locale]);
    if (!isAuthenticated) {
        return null;
    }
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement("div", { id: MY_ACCOUNT_DIV_NAME }),
        loading && react_1.default.createElement(Loading, null)));
};
const Page = () => (react_1.default.createElement(Layout_1.default, null,
    react_1.default.createElement(Container_1.default, null,
        react_1.default.createElement(SSR_1.default, { fallback: react_1.default.createElement(Loading, null) },
            react_1.default.createElement(MyAccount, null)))));
exports.default = Page;
